contracts/state-validator/src/verifications/finalize_withdrawal/last_finalized_withdrawal.rs

```diff
@@ -0,0 +1,434 @@
+use core::{ops::RangeInclusive, result::Result};
+
+use alloc::vec::Vec;
+use gw_utils::{
+    ckb_std::debug,
+    error::Error,
+    gw_common::{
+        merkle_utils::{ckb_merkle_leaf_hash, CBMTMerkleProof},
+        H256,
+    },
+    gw_types::{
+        packed::{
+            LastFinalizedWithdrawal, RawL2BlockReader, RawL2BlockWithdrawalsReader,
+            RawL2BlockWithdrawalsVecReader,
+        },
+        prelude::Unpack,
+    },
+};
+
+// Use this value, we don't need to submit prev block witness if all withdrawals are finalized
+pub const LAST_FINALIZED_WITHDRAWAL_INDEX_ALL_WITHDRAWALS: u32 = u32::MAX;
+
+#[must_use]
+pub fn check(
+    last_finalized_block_number: u64,
+    block_withdrawals_vec: &RawL2BlockWithdrawalsVecReader,
+    prev_last_finalized_withdrawal: LastFinalizedWithdrawal,
+    post_last_finalized_withdrawal: LastFinalizedWithdrawal,
+) -> Result<(), Error> {
+    debug!("check last_finalized_withdrawal");
+
+    if block_withdrawals_vec.is_empty() {
+        debug!("witness doens't have blocks");
+        return Err(Error::InvalidRollupFinalizeWithdrawalWitness);
+    }
+
+    let prev_finalized_block_number = prev_last_finalized_withdrawal.block_number().unpack();
+    let post_finalized_block_number = post_last_finalized_withdrawal.block_number().unpack();
+    debug!("prev finalized block {}", prev_finalized_block_number);
+    debug!("post finalized block {}", post_finalized_block_number);
+
+    if post_finalized_block_number < prev_finalized_block_number {
+        debug!("post block number < prev block number");
+        return Err(Error::InvalidLastFinalizedWithdrawal);
+    }
+    if post_finalized_block_number > last_finalized_block_number {
+        debug!("post block number  > last finalized block number");
+        return Err(Error::InvalidLastFinalizedWithdrawal);
+    }
+
+    let prev_last_finalized_index = LastFinalizedWithdrawalIndex::from_last_finalized_withdrawal(
+        &prev_last_finalized_withdrawal,
+    );
+    let post_last_finalized_index = LastFinalizedWithdrawalIndex::from_last_finalized_withdrawal(
+        &post_last_finalized_withdrawal,
+    );
+    debug!("prev finalized idx {:?}", prev_last_finalized_index);
+    debug!("post finalized idx {:?}", post_last_finalized_index);
+
+    // Same block rule:
+    // 1. prev last finalized withdrawal index isn't INDEX_ALL_WITHDRAWALS
+    // 2. witness block match block number
+    // 3. block should have withdrawals
+    // 4. post index must be greater than prev index
+    // 5. witness withdrawals match index range len
+    // 6. witness withdrawals have valid cbmt merkle proof
+    if post_finalized_block_number == prev_finalized_block_number {
```

Comment:
Why can't we remove the `if` branch here?

0. iterate over all blocks
1. construct a range of withdrawal indexes for each block
2. call `check_block_withdrawals`.